orderbook/src/orderbook_pool_fetcher.rs

```diff
@@ -0,0 +1,200 @@
+// - The mutex is never held while waiting on an async operation (getting pools from the node).
+// - Updating the cache is decoupled from normal pool fetches.
+// When pools are requested we mark all those pools as recently used which potentially evicts other
+// pairs from the pair lru cache. Cache misses are fetched and inserted into the cache.
+// Then when update runs the next time we request and cache all recently used pairs.
+// For the order book we only care about the "recent" state of the pools. So we can return any
+// result from the cache even if it comes from previous blocks.
+// On the other hand for the driver we need to get the pool at exact blocks which is why we keep of
+// previous blocks in the first place as we could simplify this module if it was only used by the
+// order book.
+
+impl OrderbookPoolFetcher {
+    /// number_of_blocks_to_cache: Pools for previous blocks stay in the cache this long.
+    /// number_of_pairs_to_auto_update: This number of most recently used pools get cached when the
+    ///   the current block changes.
+    pub fn new(
+        current_block: u64,
+        number_of_blocks_to_cache: u64,
+        number_of_pairs_to_auto_update: usize,
+        inner: Box<dyn PoolFetching>,
+    ) -> Self {
+        Self {
+            mutexed: Mutex::new(Mutexed::new(number_of_pairs_to_auto_update, current_block)),
+            number_of_blocks_to_cache,
+            inner,
+        }
+    }
+
+    pub async fn update_cache(&self, new_block: u64) -> Result<()> {
+        let pairs = self
+            .mutexed
+            .lock()
+            .unwrap()
+            .recently_used_pairs()
+            .collect::<HashSet<_>>();
+        let pools = self
+            .inner
+            .fetch(pairs.clone(), BlockNumber::Number(new_block.into()))
+            .await?;
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            mutexed.insert(new_block, pairs.into_iter(), &pools);
+            let oldest_to_keep = new_block.saturating_sub(self.number_of_blocks_to_cache) + 1;
+            mutexed.remove_cached_blocks_older_than(oldest_to_keep);
+            mutexed.last_update_block = new_block;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl PoolFetching for OrderbookPoolFetcher {
+    async fn fetch(&self, pairs: HashSet<TokenPair>, block: BlockNumber) -> Result<Vec<Pool>> {
+        let block = match block {
+            BlockNumber::Number(block) => Some(block.as_u64()),
+            BlockNumber::Latest => None,
+            BlockNumber::Earliest | BlockNumber::Pending => {
+                // Panic because we should never write code that does this.
+                panic!("cache does not support earliest and pending");
+            }
+        };
+
+        let mut cache_hits = Vec::new();
+        let mut cache_misses = HashSet::new();
+        let last_update_block;
+        {
+            let mutexed = self.mutexed.lock().unwrap();
```

Comment:
Not releasing the lock would not allow simultaneous reads and fetching (in case of multiple clients). My intuition leans towards only holding the lock for reading and updating the cache would perform worse.

Reply:
I should have noted that in a comment.
This is intentional and has the trade offs you mention. There is a chance that some work is wasted. The upside is that it guarantees no one has to wait for other fetches. My intuition is that it is unlikely that the same tokens are queried at the same time. And once it has happened the result is cached.
There are more complicated solutions but I would like to have the base version first before attempting those.

Reply:
> My intuition is that it is unlikely that the same tokens are queried at the same time.

Given that the base token pairs should be kept up to date with the `run_maintenance`, I also think, as you mentioned, it is more likely to have cache misses for different tokens than the same one. Also, during these cache misses, other would be stuck waiting for tokens that are already cached.

I think its better to optimize for concurrency here than for worse case.

> There are more complicated solutions but I would like to have the base version first before attempting those.

Very much agree. I would even guess that the gains would be fairly marginal from trying to sync the same fetches (as I suggested :point_up:).